vc-got.el

```diff
@@ -568,5 +568,64 @@ LIMIT limits the number of commits, optionally starting at START-REVISION."
                  (vc-got--diff file)))
               (t (error "Not implemented")))))))
 
+(defun vc-got-annotate-command (file buf &optional rev)
+  "Show annotated contents of FILE in buffer BUF. If given, use revision REV."
+  (let (process-file-side-effects)
+    (save-current-buffer
+      (set-buffer buf)
+      (apply #'vc-got--call "blame" (if rev
+                                        (list "-c" rev file)
+                                      (list file))))))
+;; captures 1=REV, 2=DATE, 3=USER
+(defconst vc-got-annotate-re
+  (concat "^[0-9]\\{1,\\}) " ; line number followed by )
+          "\\([a-z0-9]+\\) " ; SHA-1 of commit
+          "\\([0-9]\\{4\\}-[0-9]\\{2\\}-[0-9]\\{2\\}\\) "
+          "\\([^ ]\\)+ ")) ; year-mm-dd
+
+(defun vc-got-annotate-time ()
+  "Return the time of the next line of annotation at or after point.
+Value is returned as floating point fractional number of days."
+  (save-excursion
+    (beginning-of-line)
+    (when (looking-at vc-got-annotate-re)
+      (let ((str (match-string-no-properties 2)))
+        (vc-annotate-convert-time
+         (encode-time 0 0 0
+                      (string-to-number (substring str 8 10))
+                      (string-to-number (substring str 5 7))
+                      (string-to-number (substring str 0 4))))))))
+
+(defun vc-got-annotate-extract-revision-at-line ()
+  "Returns revision corresponding to the current line or nil."
+  (save-excursion
+    (beginning-of-line)
+    (when (looking-at vc-got-annotate-re)
+      (match-string-no-properties 1))))
+
+(defun vc-got-previous-revision (file rev)
+  "Return the revision number that precedes REV for FILE, or nil if no such revision exists."
+  (with-temp-buffer
+    (vc-got--log file 2 rev nil nil t)
+    (keep-lines "^commit")
+    (when (looking-at "^commit \\([a-z0-9]+\\)")
+      (match-string-no-properties 1))))
+
+(defun vc-got-next-revision (file rev)
+  "Return the revision number that follows REV for FILE, or nil
+  if no such revision exists."
+  (with-temp-buffer
+    (vc-got--log file nil nil rev)
+    (keep-lines "^commit")
+    (goto-char (point-max))
+    (beginning-of-line)
+    (previous-line) ;; return from empty line to last actual commit
+    (when (looking-at  "^commit \\([a-z0-9]+\\)$")
```

Comment:
(sorry for the spam)

the `$` shouldn't be there.  I was debugging this and realised that the point is looking at

```
commit 8da4c01ab1701f49af2bb11b89698b2b3f839f1b (master, main, github/master, origin/master)
```

Reply:
Yeah, it was there originally to omit that commit as I think that points to to top level commit. But I think the "string= rev ..." comparison checks that so the $ is not needed.